File: server/modules/keybindings/store.ts
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { homedir } from "node:os";
import { join, dirname } from "node:path";
import type { KeybindingsMap } from "../../schemas/api.js";

export const KEYBINDING_OPTIONS_PATH = join(homedir(), ".workbench", "keybindings.json");

/** macOS US: Option+1…9 → ¡ ™ £ ¢ ∞ § ¶ • ª */
const MAC_OPTION_TAB_CHARS = ["¡", "™", "£", "¢", "∞", "§", "¶", "•", "ª"] as const;

export const KEYBINDING_OPTIONS: KeybindingsMap = {
  "terminal.newTerminal": "Ctrl+Shift+n",
  "panel.explorer": "Ctrl+Shift+e",
  "panel.git": "Ctrl+Shift+g",
  "settings.open": "Ctrl+Shift+,",
  "terminal.tab.1": `Option+${MAC_OPTION_TAB_CHARS[0]}`,
  "terminal.tab.2": `Option+${MAC_OPTION_TAB_CHARS[1]}`,
  "terminal.tab.3": `Option+${MAC_OPTION_TAB_CHARS[2]}`,
  "terminal.tab.4": `Option+${MAC_OPTION_TAB_CHARS[3]}`,
  "terminal.tab.5": `Option+${MAC_OPTION_TAB_CHARS[4]}`,
  "terminal.tab.6": `Option+${MAC_OPTION_TAB_CHARS[5]}`,
  "terminal.tab.7": `Option+${MAC_OPTION_TAB_CHARS[6]}`,
  "terminal.tab.8": `Option+${MAC_OPTION_TAB_CHARS[7]}`,
  "terminal.tab.9": `Option+${MAC_OPTION_TAB_CHARS[8]}`,
};

export async function getKeybindings(
  filePath = KEYBINDING_OPTIONS_PATH,
): Promise<KeybindingsMap> {
  try {
    const raw = await readFile(filePath, "utf8");
    const parsed = JSON.parse(raw) as Partial<KeybindingsMap>;
    return { ...KEYBINDING_OPTIONS, ...parsed };
  } catch {
    return { ...KEYBINDING_OPTIONS };
  }
}

export async function putKeybindings(
  map: KeybindingsMap,
  filePath = KEYBINDING_OPTIONS_PATH,
): Promise<void> {
  await mkdir(dirname(filePath), { recursive: true });
  await writeFile(filePath, JSON.stringify(map, null, 2), "utf8");
}

File: server-rs/src/keybindings/store.rs
use crate::db::get_data_dir;
use crate::error::{AppError, AppResult};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

use std::sync::LazyLock;

pub type KeybindingsMap = HashMap<String, String>;

pub static KEYBINDING_OPTIONS: LazyLock<KeybindingsMap> =
    LazyLock::new(default_keybinding_options);

pub static KEYBINDING_OPTIONS_PATH: LazyLock<PathBuf> =
    LazyLock::new(|| get_data_dir().join("keybindings.json"));

pub fn keybinding_options_path() -> PathBuf {
    KEYBINDING_OPTIONS_PATH.clone()
}

/// macOS US: Option+1…9 → ¡ ™ £ ¢ ∞ § ¶ • ª
const MAC_OPTION_TAB_CHARS: [&str; 9] = ["¡", "™", "£", "¢", "∞", "§", "¶", "•", "ª"];

pub fn default_keybinding_options() -> KeybindingsMap {
    let mut map = HashMap::new();
    map.insert("terminal.newTerminal".to_string(), "Ctrl+Shift+n".to_string());
    map.insert("panel.explorer".to_string(), "Ctrl+Shift+e".to_string());
    map.insert("panel.git".to_string(), "Ctrl+Shift+g".to_string());
    map.insert("contextQueue.invoke".to_string(), "Ctrl+l".to_string());
    map.insert("settings.open".to_string(), "Ctrl+Shift+,".to_string());
    for (i, ch) in MAC_OPTION_TAB_CHARS.iter().enumerate() {
        map.insert(format!("terminal.tab.{}", i + 1), format!("Option+{ch}"));
    }
    map
}

pub fn get_keybindings(file_path: Option<&PathBuf>) -> AppResult<KeybindingsMap> {
    let path = file_path.cloned().unwrap_or_else(keybinding_options_path);
    let defaults = default_keybinding_options();

    let raw = match fs::read_to_string(&path) {
        Ok(raw) => raw,
        Err(_) => return Ok(defaults),
    };

    let parsed: HashMap<String, String> = match serde_json::from_str(&raw) {
        Ok(parsed) => parsed,
        Err(_) => return Ok(defaults),
    };

    let mut merged = defaults;
    merged.extend(parsed);
    Ok(merged)
}

pub fn put_keybindings(map: &KeybindingsMap, file_path: Option<&PathBuf>) -> AppResult<()> {
    let path = file_path.cloned().unwrap_or_else(keybinding_options_path);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| AppError::Internal(e.to_string()))?;
    }
    let json = serde_json::to_string_pretty(map)
        .map_err(|e| AppError::Internal(e.to_string()))?;
    fs::write(&path, json).map_err(|e| AppError::Internal(e.to_string()))?;
    Ok(())
}
